blog/_posts/2022-08-08-managing-tunnels-with-argocd.md

```diff
@@ -0,0 +1,236 @@
+
+```
+# Example umbrella chart
+├── Chart.yaml
+└── values-postgresql-customer1.yaml
+```
+
+In the `Chart.yaml` of the umbrella chart we specify the `inlets-pro` chart as a dependency. We can than add multiple values files, one for each server that has to be created. The different values files in the chart can be referenced later by the Argo CD apps for each tunnel server.
+```yaml
+apiVersion: v2
+name: inlets
+description: Helm chart for an inlets-pro TCP server
+type: application
+version: 0.1.0
+appVersion: "0.1.0"
+dependencies:
+  - name: inlets-pro
+    version: 0.4.0
+    repository: https://inlets.github.io/inlets-pro/charts
+```
+
+The `values-postgresql-customer1.yaml` sets the configuration values for the first tunnel server:
+```yaml
+inlets-pro:
+  tokenSecretName: postgresql-customer1-token
+  ingress:
+    domain: postgresql-customer1.example.com
+
+  dataPlane:
+    type: ClusterIP
+    ports:
+    - targetPort: 5432
+      protocol: TCP
+      name: postgresql
+      port: 5432
+```
+
+We set the address, in this case `postgresql-customer1.example.com`, that we want our inlets clients to connect to in the `ingress.domain` field. When the chart is deployed a Certificate is created and cert-manager will request a TLS certificate from Let’s Encrypt using a HTTP01 challenge.
+
+In the `tokenSecretName` field we set the name of secret containing the token that we want to use for this server. In the next section we will see how to create these these secrets and make them available in the cluster.
+
+## Manage tokens for your tunnels
+Each inlets tunnel server requires a unique token. This token needs to be stored as a secret in the cluster so that it can be accessed by the inlets server. Argo CD is un-opinionated about how secrets are managed. Wether you want to use Vault, Helm Secrets or SOPS, to name a few, they can all be set up to work with ArgoCD. We will be using [Bitnami Sealed Secrets](https://github.com/bitnami-labs/sealed-secrets).
+
+Install sealed-secrets and the `kubeseal` cli:
+```bash
+# Install sealed-secrets
+arkade install sealed-secrets
+
+# Install cli
+arkade get kubeseal
+```
+
+The `kubeseal` cli will let us encrypt a Kubernetes Secret using asymmetric encryption. It creates a SealedSecret CRD that can safely be stored in version control. Once a SealedSecret is deployed to the cluster, the Sealed Secrets controller will decrypt it and deploy a normal Kubernetes Secret.
+
+Create a token for your inlets-pro server:
+```bash
+# Generate a random token
+export TOKEN=$(head -c 16 /dev/random | shasum|cut -d" " -f1)
+
+# Create a yaml-encoded Secret without adding it to the cluster
+kubectl create secret generic postgresql-customer1-token \
+  --dry-run=client \
+  --from-literal=token=$TOKEN \
+  -o yaml \
+  > postgresql-customer1-token-secret.yaml
+```
+
+Use the `kubeseal` cli to create a `SealedSecret` from the `Secret` we generated with the previous command.
+```bash
+kubeseal \
```

Comment:
I prefer exporting the public key in a separate stage and then running kubeseal using it as an argument.

Why?

The command above needs high levels of RBAC permissions.

But anyone can run kubeseal with the public key, without any cluster access.

These are the old instructions from OFC: https://github.com/openfaas/openfaas-cloud/blob/master/docs/README.md#sealedsecret-support

Reply:
I updated the instructions to first export the public key and run kubeseal using it as an argument.